src/cobra/core/reaction.py

```diff
@@ -590,13 +590,14 @@ def genes(self) -> FrozenSet:
         """
         return frozenset(self._genes)
 
-    def _update_genes_from_gpr(self, new_gene_names: Optional[Set] = None) -> None:
+    def update_genes_from_gpr(self, new_gene_names: Optional[Set] = None) -> None:
```

Comment:
Shouldn't `new_gene_names` be a dict? Wouldn't you need the old name and new name for each gene here to rename them?

Reply:
No. This is a bit confusing, because it is a long revamp of the gene_reaction_rule setter.
Before I used GPR() as class, gene_reaction_rule setter would parse the gene_rule, and then use split and such like to get the gene_names. Then it would copy the old_names, and compare it to gene_names.
If the reaction had a model, it would either point to genes in the model, or add new genes that were not in the model. If the reaction didn't have a model, it would just create new genes as Gene(gene_id). Then genes in old_genes, not in new_genes would be removed from the reaction (and possibly the model).

update_genes_from_gpr() tries to do the roughly the same thing - update reaction._genes with new_genes and remove old genes. Now that you mention it, new_gene_names should probably NOT be a parameter, since it is taken from GPR() (and then if new_gene_names is None: will always be true, so it can be removed).

Reply:
Yeah, I agree that the gene names should be taken from the GPR only here and not via an argument. I think the behavior looks good though. I did not see it deleting genes here. However it tries to add genes even if they already exist in the model genes it seems so that may be skipped probably.

Reply:
It adds the genes to the current model if the reaction has a model, and these genes exist, but aren't in the reaction. Unless I misunderstood you?